### PR TITLE
Allow setting Player Notes and more interactive player objects

### LINKED ISSUE #2539: Add the ability to record (edit) a note for each player
The notes are stored in a local file. If there is a note for the player, display it as an icon next to the player's name. When you hover over the icon, the text is displayed

### LINKED ISSUE #2592: Show the user that it is possible to interact with the player card

### PR DESCRIPTION
**Pull request by Marc-Spector**
Closes #2539 
Closes #2592

This is still a draft version. I will add UI items in the future Tips and suggestions are welcome.

<details>
<summary>Demo</summary>

demo
</details>

<details>
<summary> More UI interaction with the player card</summary>

demo1
</details>

### CODE REVIEW

File: src/main/java/com/faforever/client/player/PlayerNoteService.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  private final ObjectMapper objectMapper;
+  private final EventBus eventBus;
+
+  private Path notesPath;
+  private Map<Integer, String> notes;
+
+  @Override
+  public void afterPropertiesSet() throws Exception {
+    eventBus.register(this);
+    notesPath = preferencesService.getPreferencesDirectory().resolve(PLAYER_NOTES_FILE_NAME);
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/player/PlayerNoteService.java]: I think it might make more sense to just store the notes in the preferences object itself as an observable map, then you can also add a listener onto the map to add the note to the player. I don't see any advantages to having it in a separate file.

**Comment by Marc-Spector** [comment]
It is also necessary to set a limit on the number of notes. I think 1500 is reasonable.

File: src/main/java/com/faforever/client/player/PlayerNoteService.java (hunk 1 above)
Comment (Marc-Spector) [reply src/main/java/com/faforever/client/player/PlayerNoteService.java]: Ok, I will try.

**Comment by Sheikah45** [comment]
Why do we need a limit on the number of notes? Since it is all stored in local memory the only danger is that people fill up all their ram with notes but given it is all text that would be pretty hard to do.

**Comment by Marc-Spector** [comment]
> Why do we need a limit on the number of notes? Since it is all stored in local memory the only danger is that people fill up all their ram with notes but given it is all text that would be pretty hard to do.

Okay. But still, do not underestimate people. 😄

**Comment by Sheikah45** [comment]
> > Why do we need a limit on the number of notes? Since it is all stored in local memory the only danger is that people fill up all their ram with notes but given it is all text that would be pretty hard to do.
> 
> Okay. But still, do not underestimate people. 😄 

I am not ha, but at the end of the day it can have no affect on other users or the server so see little reason to limit it.

File: src/main/java/com/faforever/client/fx/contextmenu/AddPlayerNoteMenuItem.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  @Override
+  protected boolean isItemVisible() {
+    return object != null && !playerService.containsNote(object);
+  }
+
+  @Override
+  protected String getItemText() {
+    return i18n.get("chat.userContext.addNote");
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/fx/contextmenu/AddPlayerNoteMenuItem.java]: Why are this and EditPlayerNote two separate objects? Why not just have a single class that gives different text depending on whether there is a player note that already exists for this player, as otherwise they are identical.

File: src/main/java/com/faforever/client/fx/contextmenu/EditPlayerNoteMenuItem.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  @Override
+  protected boolean isItemVisible() {
+    return object != null && playerService.containsNote(object);
+  }
+
+  @Override
+  protected String getItemText() {
+    return i18n.get("chat.userContext.editNote");
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/fx/contextmenu/EditPlayerNoteMenuItem.java]: See comment on addPlayerNote

File: src/main/java/com/faforever/client/player/PlayerService.java [hunk 4]
```diff
@@ -194,6 +210,26 @@ void createOrUpdatePlayerForPlayerInfo(@NotNull com.faforever.commons.lobby.Play
     resetIdleTime(player);
   }
 
+  public boolean containsNote(PlayerBean player) {
+    return playerNotes.containsKey(player.getId());
+  }
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/player/PlayerService.java]: Couldn't we just check if player.getNote() is blank since the player note should be set on all playerBean initialization?

File: src/main/java/com/faforever/client/player/PlayerService.java [hunk 5]
```diff
@@ -194,6 +210,26 @@ void createOrUpdatePlayerForPlayerInfo(@NotNull com.faforever.commons.lobby.Play
...
+  public void updateNote(PlayerBean player, String text) {
+    int id = player.getId();
+
+    if (StringUtils.isBlank(text)) {
+      removeNote(player);
+    } else if (playerNotes.containsKey(id)) {
+      playerNotes.replace(id, text);
+    } else {
+      playerNotes.put(id, text);
+    }
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/player/PlayerService.java]: Why is this necessary seems like we don't need the else if

File: src/main/java/com/faforever/client/preferences/Preferences.java [hunk 6]
```diff
@@ -89,6 +89,7 @@
   IntegerProperty cacheLifeTimeInDays = new SimpleIntegerProperty(30);
   BooleanProperty gameDataCacheActivated = new SimpleBooleanProperty(false);
   BooleanProperty debugLogEnabled = new SimpleBooleanProperty(false);
+  MapProperty<Integer, String> playerNotes = new SimpleMapProperty<>(observableHashMap()); // key - player ID
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/preferences/Preferences.java]: Rather than the comment just name the variable `notesByPlayerId` like the other maps are named.

Also this should go into one of the sub preference objects or a new one. I am trying to move away from having things in the top level object

**Review by Sheikah45 — CHANGES_REQUESTED** [review]

File: src/main/java/com/faforever/client/fx/contextmenu/DoPlayerNoteMenuItem.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+@RequiredArgsConstructor
+public class DoPlayerNoteMenuItem extends AbstractMenuItem<PlayerBean> {
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/fx/contextmenu/DoPlayerNoteMenuItem.java]: Can just stay as EditPlayerNoteMenuItem

File: src/main/java/com/faforever/client/player/PlayerNoteController.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    textArea.setTextFormatter(formatter);
+    JavaFxUtil.addListener(textArea.lengthProperty(), observable ->
+        charactersCountLabel.setText(String.format("%d / %d", textArea.getLength(), maxSymbolLimit)));
+  }
+
+  private boolean isSymbolLimitReached(Change change) {
+    return change.getControlNewText().length() > maxSymbolLimit;
+  }
+
+  private boolean isMultipleSeparators(Change change) {
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/player/PlayerNoteController.java]:
```suggestion
  private boolean hasMultipleSeparators(Change change) {
```

File: src/main/java/com/faforever/client/player/PlayerNoteController.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  public void okButtonClicked() {
+    playerService.updateNote(player, textArea.getText());
+    cancelButton.fire();
+  }
+
+  public void setCloseButtonAction(EventHandler<ActionEvent> handler) {
+    cancelButton.setOnAction(handler);
+  }
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/player/PlayerNoteController.java]: When would this note be dialog.close()?

**Review by Sheikah45 — CHANGES_REQUESTED** [review]
Almost there although tests are failing

File: src/main/java/com/faforever/client/player/PlayerNoteController.java (hunk 9 above)
Comment (Marc-Spector) [reply src/main/java/com/faforever/client/player/PlayerNoteController.java]: Sorry, I don't understand what you want.

Comment (Sheikah45) [reply src/main/java/com/faforever/client/player/PlayerNoteController.java]: Sorry there was a typo. Although the main question is why does this function exist. The onClose action is always the same and I am not sure of a situation where it would be different.

Comment (Marc-Spector) [reply src/main/java/com/faforever/client/player/PlayerNoteController.java]: Fixed.

**Review by Sheikah45 — APPROVED** [review]
